File: DotNet/Exercicios1/Combiner.cs
using System;
using System.Collections.Generic;
using System.Threading;
using Utils;

namespace Exercicios1 {
    public class Combiner<L, R> where L : class
                                where R : class {

        private object monitor = new object();      // the monitor object

        private LinkedList<Offer> lOffers;          // pending L offers
        private LinkedList<Offer> rOffers;          // pending R offers
        private LinkedList<Pair> pairs;             // pairs queue
        private LinkedList<object> waiters;         // waiters queue
    

        private class Offer {
            private L offerL;
            private R offerR;

            public Offer(L lval) { offerL = lval;  }
            public Offer(R rval) { offerR = rval;  }

            public L OfferL{
                get { return offerL; }
            }

            public R OfferR {
                get { return offerR; }
            }
        }

      

        public Combiner() {
            pairs = new LinkedList<Pair>();
            waiters = new LinkedList<object>();
            rOffers = new LinkedList<Offer>();
            lOffers = new LinkedList<Offer>();
        }


        public class Pair {
            public readonly L left;
            public readonly R right;

            internal Pair(L left, R right) {
                this.left = left; this.right = right;
            }
            
           
        }

        public void PutLeft(L left) {
            lock (monitor) {
                // check if the there are rOffers!
                if (rOffers.Count > 0) {
                    Offer o = rOffers.First.Value;
                    rOffers.RemoveFirst();
                    pairs.AddLast(new Pair(left, o.OfferR));
                    if (waiters.Count > 0) {
                        Monitor.PulseAll(monitor);  // the front requested pair is completed!
                    }         
                }
                else {
                    lOffers.AddLast(new Offer(left));
                }
            }
        }

        public void PutRight(R right) {
            lock (monitor) {
                // check if the there are lOffers!
                if (lOffers.Count > 0) {
                    Offer o = lOffers.First.Value;
                    lOffers.RemoveFirst();
                    pairs.AddLast(new Pair(o.OfferL, right));
                    if (waiters.Count > 0) {
                        Monitor.PulseAll(monitor);  // the front requested pair is completed!
                    }
                }
                else {
                    rOffers.AddLast(new Offer(right));
                }
            }
        }

        public Pair Take(int timeout) // throws ThreadInterruptedException
        {
            lock (monitor) {
                // fast path
                if (waiters.Count > 0 /* avoid barging */ && pairs.Count > 0) {
                    var pair = pairs.First.Value;
                    pairs.RemoveFirst();
                    return pair;
                }
                if (timeout == 0) {
                    return null;
                }
                // prepare wait
                var req = waiters.AddLast((object)null);
                do {
                    try {
                        int refTime = Environment.TickCount;
                        Monitor.Wait(monitor, timeout);
                        if (waiters.First == req && pairs.Count > 0) {
                            var pair = pairs.First.Value;
                            pairs.RemoveFirst();
                            return pair;
                        }
                        timeout = SynchUtils.RemainingTimeout(refTime, timeout);
                        if (timeout == 0) {
                            // abort wait
                            waiters.Remove(req);
                            return null;
                        }
                    }
                    catch (ThreadInterruptedException) {
                        // abort wait
                        waiters.Remove(req);
                        throw;
                    }
                }
                while (true);
            }
        }
    }
}

File: Java/src/exercicios1/Combiner.java
package exercicios1;

import java.util.LinkedList;
import java.util.List;

import static utils.SynchUtils.INFINITE;
import static utils.SynchUtils.remainingTimeout;

/**
 * Created by jmartins on 09/10/2017.
 */
public class Combiner<L,R>  {
    private Object monitor = new Object(); // the monitor object

    private List<L> lOffers;    // queue for L offers
    private List<R> rOffers;    // queue for R offers

    private List<Object> waiters;   // pair Waiters
    private List<Pair> pairs;       // completed pairs

    public Combiner() {
        lOffers = new LinkedList<>();
        rOffers = new LinkedList<>();
        waiters = new LinkedList<>();
        pairs =   new LinkedList<>();
    }

    /**
     * Inner class for pair holding
     */
    public class Pair {
        public final L left;
        public final R right;

        private Pair(L left, R right) {
            this.left = left; this.right = right;
        }

    }

    public void PutLeft(L left) {

        synchronized(monitor) {
            // check if the there are rOffers!
            if (rOffers.size() > 0) {
                R r  = rOffers.remove(0);
                pairs.add(new Pair(left, r));
                if (waiters.size() > 0) {
                    monitor.notifyAll(); // the front requested pair is completed!

                }
            }
            else {
                lOffers.add(left);
            }
        }
    }

    public void PutRight(R right) {
        synchronized(monitor) {
            // check if the there are lOffers!
            if (lOffers.size() > 0) {
                L l = lOffers.remove(0);
                pairs.add(new Pair(l, right));
                if (waiters.size() > 0) {
                    monitor.notifyAll(); // the front requested pair is completed!
                    return;
                }
                // if not, insert a new pair on the list

            }
            else {
                rOffers.add(right);
            }
        }
    }


    public Pair Take(long timeout)throws  InterruptedException
    {
        synchronized(monitor)
        {
            // fast path
            if (waiters.size() == 0 /* avoid barging */ && pairs.size() > 0 )
                return  pairs.remove(0);
            if (timeout == 0)
                return null;

            // prepare wait
            Object req = new Object();
            waiters.add(req);
            do
            {
                try {
                    long refTime = System.currentTimeMillis();
                    if (timeout == INFINITE)
                        monitor.wait();
                    else
                        monitor.wait(timeout);
                    if (waiters.get(0) == req && pairs.size() > 0)
                        return pairs.remove(0);

                    timeout =  remainingTimeout(refTime, timeout);
                    if (timeout == 0)  {
                        // abort wait
                        waiters.remove(req);
                        return null;
                    }
                }
                catch( InterruptedException e) {
                    // abort wait
                    waiters.remove(req);
                    throw e;
                }
            }
            while (true);
        }
    }
}